File: other/minPathSum.py
"""
给定一个 n * m 的矩阵 a，从左上角开始每次只能向右或者向下走，最后到达右下角的位置，
路径上所有的数字累加起来就是路径和，输出所有的路径中最小的路径和。
"""
class Solution:
    def minPathSum(self, arr, n, m):
        return self.process(arr, 0, 0, n, m)

    def process(self, arr, x, y, n, m):
        if x == m - 1 and y == n - 1:
            return arr[x][y]

        cur = arr[x][y]

        if x == m - 1:
            return cur + self.process(arr, x, y+1, n, m)

        if y == n - 1:
            return cur + self.process(arr, x+1, y, n, m)

        return cur + min(self.process(arr, x+1, y, n, m), self.process(arr, x, y+1, n, m))


arr = [[1,3,5,9],[8,1,3,4],[5,0,6,1],[8,8,4,0]]
n, m = 4, 4
solution = Solution()
print(solution.minPathSum(arr, n, m))

File: leetcode_hot_100/minPathSum.cpp
// hot100-lc 64. 最小路径和

class Solution {
public:
    int minPathSum(std::vector<std::vector<int>>& grid)
    {
        int rows = (int)grid.size();
        int cols = (int)grid[0].size();
        int dp[rows][cols];
        dp[0][0] = grid[0][0];
        for (int i = 1; i < rows; i++)
        {
            dp[i][0] = dp[i-1][0] + grid[i][0];
        }
        for (int j = 1; j < cols; j++)
        {
            dp[0][j] = dp[0][j-1] + grid[0][j];
        }
        for (int i = 1; i < rows; i++)
        {
            for (int j = 1; j < cols; j++)
            {
                dp[i][j] = std::min(dp[i-1][j] + grid[i][j], dp[i][j-1] + grid[i][j]);
            }
        }
        return dp[rows-1][cols-1];
    }
};
